### PR TITLE
Delay KV bulk delete

### LINKED ISSUE #783: Bug: Downtime on Site update
When we publish a `site` worker, often the worker and its accompanying asset manifest make it to the edge ahead of the new KV keys, and/or the "old" KV keys are removed before the worker is updated. This causes a window of downtime when a page is updated. We need to somehow delay removing "old" assets from KV until everything is live. There are a few ways to implement this, will update this issue with possible strategies soon.

### PR DESCRIPTION
**Pull request by ashleymichal**
## Overview
Fixes #783 

Previously, we would run a full KV namespace sync BEFORE uploading the worker script to the edge. This caused some unacceptable downtime for updated assets, as the Worker would continue  handling requests to paths with outdated KV keys. This PR does a few things in service of re-ordering these operations, but the bottom line is that now on Wrangler publish, when the project is a Workers Site, we:

1. validate the wrangler.toml
2. run `wrangler build`
3. query the site's KV namespace for existing keys and diff them against the local site `bucket` to build the list of files to upload, the list of keys to delete, and the asset manifest to be uploaded with the Worker script.
4. execute the bulk file upload
5. upload the worker with its new asset manifest
6. execute the bulk file delete

This has the added benefit of NOT deleting old values in the case where the Worker upload fails, since the first KV operation is additive.

## Things this is not
This PR introduces a little more verbosity into preview/upload and publish, however there is considerable overlap between those operations already, but I'd like to address those in a separate PR in order to expedite this fix.

I'm also interested in optimizing the preview command's behavior (refrain from deleting until a `ctrl-C` or something like that) but not here.

## 💅 
There's a few house-keeping changes here, though more remain which I'll include as inline comments:

* we now pass a set of KeyValuePairs to kv::bucket::upload, rather than a bucket.
* some house-keeping in the publish command, mostly renaming and some refactors
* ~remove unused/unsupported `bucket` attribute from KvNamespace struct~

### CODE REVIEW

File: src/commands/kv/bucket/mod.rs [hunk 1]
```diff
@@ -32,43 +33,53 @@ pub fn directory_keys_values(
...
-            asset_manifest.insert(url_safe_path, key);
+                    asset_manifest.insert(url_safe_path, key);
+                }
+            }
+            Ok((upload_vec, asset_manifest))
+        }
+        Ok(_file_type) => {
+            // any other file types (files, symlinks)
+            // TODO: return an error type here, like NotADirectoryError
+            Err(format_err!("site.bucket must be a directory"))
```
Comment (EverlastingBugstopper) [inline src/commands/kv/bucket/mod.rs]: Can we get a more actionable error message here that includes the value of what they have set as their bucket and direct them to their `wrangler.toml`? "site.bucket" is a bit opaque.

File: src/commands/publish.rs [hunk 2]
```diff
@@ -57,73 +80,57 @@ fn warn_site_incompatible_route(deploy_config: &DeployConfig) {
...
+            "{} bucket directory \"{}\" does not exist",
+            emoji::WARN,
+            path.display()
+        )
+    } else if !path.is_dir() {
+        failure::bail!(
+            "{} bucket \"{}\" is not a directory",
+            emoji::WARN,
+            path.display()
+        )
```
Comment (EverlastingBugstopper) [inline src/commands/publish.rs]: these failures are not related to if it is a root bucket, should probably rename the function 

(worth noting that it's also invalid to have bucket be `dist` for webpack types but we can leave that for another PR)

File: src/commands/publish.rs [hunk 3]
```diff
@@ -57,73 +80,57 @@ fn warn_site_incompatible_route(deploy_config: &DeployConfig) {
...
-            }
-        }
+// We don't want folks setting their bucket to the top level directory,
+// which is where wrangler commands are always called from.
+pub fn assert_non_root_bucket(bucket: &PathBuf) -> Result<(), failure::Error> {
+    // TODO: this should really use a convenience function for "Wrangler Project Root"
+    let current_dir = env::current_dir()?;
+    if bucket.as_os_str() == current_dir {
+        failure::bail!(
+            "{} You need to specify a bucket directory in your wrangler.toml",
```
Comment (EverlastingBugstopper) [inline src/commands/publish.rs]: this seems like the wrong error message

**Comment by ashleymichal** [comment]
update: i'm not  necessarily gunning to get this into 1.8.1; i'm happy to wait on 1.9.0. one of the refactors here removes the `bucket` attribute from kv-namespaces, which from what i could tell was never supported, but i want to do a little extra testing to confirm that there isn't any support here so i don't break unsuspecting crafty users.

**Comment by AngusMorton** [comment]
> update: i'm not necessarily gunning to get this into 1.8.1; i'm happy to wait on 1.9.0. one of the refactors here removes the `bucket` attribute from kv-namespaces, which from what i could tell was never supported, but i want to do a little extra testing to confirm that there isn't any support here so i don't break unsuspecting crafty users.

Is there an ETA for 1.9.0 or for this to be released (even as an alpha/RC)?

**Comment by ashleymichal** [comment]
> Is there an ETA for 1.9.0 or for this to be released (even as an alpha/RC)?

We'll be putting this out on its own as a point release (1.8.2)

**Pull request merged**